### ISSUE TITLE
Tpetra: 'Class with virtual base class and delegating constructor' error under IBM xlC

### ISSUE THREAD

**Issue by djfitzgerald** [trilinos/Trilinos#2938]
@trilinos/tpetra

When attempting to unit test a fix I've been working on for #2781, the IBM XL compiler fails at packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp:651 with error: `1540-2206 Class with virtual base class and delegating constructor is not supported.` The failing class definition is as follows:
```
  template <class LocalOrdinal, class GlobalOrdinal, class Node>
  CrsGraph<LocalOrdinal, GlobalOrdinal, Node>::
  CrsGraph (const Teuchos::RCP<const map_type>& rowMap,
            const Teuchos::RCP<const map_type>& colMap,
            const local_graph_type& k_local_graph_,
            const Teuchos::RCP<Teuchos::ParameterList>& params)
    : CrsGraph (k_local_graph_,
                rowMap,
                colMap,
                Teuchos::null,
                Teuchos::null,
                params)
{}
```

It looks like this area of code was modified in the recent past, by https://github.com/djfitzgerald/Trilinos/commit/c5fd93a0756df81cb0cda375f135de968372f634 and https://github.com/djfitzgerald/Trilinos/commit/888258fc088fa5276137ce4655a3707bb54fd748.  Because I know very little about C++, I'd rather not be the person to attempt fixing it.  This is preventing the UT of my fix for #2781.

I am building Trilinos from Spack, using a Spack repository to recognize a fake Trilinos "version 12.13.1" that points to a working branch in my fork of this GitHub repository. From the hacked Spack package.py file for my local Trilinos install:
```
version('12.13.1', git='https://github.com/basicmanfitz/Trilinos.git', branch='fix-2781')
```
My Spack install command is:
```
spack install --test=root trilinos@12.13.1%xl_r ~pnetcdf ^netlib-lapack+external-blas ^netcdf%gcc ^m4%gcc
```
And the pertinent text from my Spack build log:
```
==> Installing trilinos
==> Trying to clone git repository: https://github.com/djfitzgerald/Trilinos.git on branch fix-2781
==> No checksum needed when fetching with git
==> Already staged trilinos-12.13.1-4kl2xjobtvkeiimmeunht447jfdolep2 in /u/danfitz/spack/var/spack/stage/trilinos-12.13.1-4kl2xjobtvkeiimmeunht447jfdolep2
==> Applied patch xlf_seacas.patch
==> Building trilinos [CMakePackage]
==> Executing phase: 'cmake'
==> Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j176'

5 errors found in build log:
     8539    cd /tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/ifpack/src && /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/cmak
             e-3.11.3-4joy3z55k5hm6j75ssjougditfpsjh5j/bin/cmake -E cmake_link_script CMakeFiles/ifpack.dir/link.txt --verbose=1
     8540    /opt/ibm/spectrum_mpi/bin/mpixlC  -DMUMPS_5_0 -std=gnu++11 -g  -qmkshrobj -Wl,-soname,libifpack.so.12 -o libifpack.so.12.13 CMakeFiles/ifpack.dir/Ifpack
             _IlukGraph.cpp.o CMakeFiles/ifpack.dir/Ifpack_CrsRiluk.cpp.o CMakeFiles/ifpack.dir/Ifpack_ILU.cpp.o CMakeFiles/ifpack.dir/Ifpack_IKLU.cpp.o CMakeFiles/i
             fpack.dir/Ifpack_IKLU_Utils.cpp.o CMakeFiles/ifpack.dir/Ifpack_ILUT.cpp.o CMakeFiles/ifpack.dir/Ifpack_ICT.cpp.o CMakeFiles/ifpack.dir/Ifpack_IC.cpp.o C
             MakeFiles/ifpack.dir/Ifpack_IC_Utils.cpp.o CMakeFiles/ifpack.dir/Ifpack_OverlapGraph.cpp.o CMakeFiles/ifpack.dir/Ifpack_MultiListSort.c.o CMakeFiles/ifp
             ack.dir/ifp_parameters.cpp.o CMakeFiles/ifpack.dir/Ifpack_DiagPreconditioner.cpp.o CMakeFiles/ifpack.dir/Ifpack_PointRelaxation.cpp.o CMakeFiles/ifpack.
             dir/Ifpack_Condest.cpp.o CMakeFiles/ifpack.dir/Ifpack.cpp.o CMakeFiles/ifpack.dir/Ifpack_DenseContainer.cpp.o CMakeFiles/ifpack.dir/Ifpack_TriDiContaine
             r.cpp.o CMakeFiles/ifpack.dir/Ifpack_SerialTriDiMatrix.cpp.o CMakeFiles/ifpack.dir/Ifpack_SerialTriDiSolver.cpp.o CMakeFiles/ifpack.dir/Ifpack_LocalFilt
             er.cpp.o CMakeFiles/ifpack.dir/Ifpack_NodeFilter.cpp.o CMakeFiles/ifpack.dir/Ifpack_SingletonFilter.cpp.o CMakeFiles/ifpack.dir/Ifpack_DropFilter.cpp.o 
             CMakeFiles/ifpack.dir/Ifpack_SparsityFilter.cpp.o CMakeFiles/ifpack.dir/Ifpack_ReorderFilter.cpp.o CMakeFiles/ifpack.dir/Ifpack_DiagonalFilter.cpp.o CMa
             keFiles/ifpack.dir/Ifpack_Graph_Epetra_RowMatrix.cpp.o CMakeFiles/ifpack.dir/Ifpack_Graph_Epetra_CrsGraph.cpp.o CMakeFiles/ifpack.dir/Ifpack_Utils.cpp.o
              CMakeFiles/ifpack.dir/Ifpack_SPARSKIT.cpp.o CMakeFiles/ifpack.dir/Ifpack_Chebyshev.cpp.o CMakeFiles/ifpack.dir/Ifpack_Polynomial.cpp.o CMakeFiles/ifpac
             k.dir/Ifpack_Krylov.cpp.o CMakeFiles/ifpack.dir/Ifpack_OverlappingPartitioner.cpp.o CMakeFiles/ifpack.dir/Ifpack_LinePartitioner.cpp.o CMakeFiles/ifpack
             .dir/Ifpack_LinearPartitioner.cpp.o CMakeFiles/ifpack.dir/Ifpack_GreedyPartitioner.cpp.o CMakeFiles/ifpack.dir/Ifpack_METISPartitioner.cpp.o CMakeFiles/
             ifpack.dir/Ifpack_EquationPartitioner.cpp.o CMakeFiles/ifpack.dir/Ifpack_UserPartitioner.cpp.o CMakeFiles/ifpack.dir/Ifpack_RCMReordering.cpp.o CMakeFil
             es/ifpack.dir/Ifpack_METISReordering.cpp.o CMakeFiles/ifpack.dir/Ifpack_OverlappingRowMatrix.cpp.o CMakeFiles/ifpack.dir/Ifpack_ValidParameters.cpp.o CM
             akeFiles/ifpack.dir/Ifpack_HIPS.cpp.o CMakeFiles/ifpack.dir/Ifpack_Hypre.cpp.o CMakeFiles/ifpack.dir/Ifpack_Euclid.cpp.o CMakeFiles/ifpack.dir/Ifpack_SI
             LU.cpp.o CMakeFiles/ifpack.dir/Ifpack_IHSS.cpp.o CMakeFiles/ifpack.dir/Ifpack_SORa.cpp.o CMakeFiles/ifpack.dir/Ifpack_AMDReordering.cpp.o CMakeFiles/ifp
             ack.dir/Ifpack_Amesos.cpp.o -Wl,-rpath,/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/amesos/src:/tmp/danfitz/spack-stage/spa
             ck-stage-AO_5dB/Trilinos/spack-build/packages/aztecoo/src:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/epetraext/src:/u/dan
             fitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/hypre-2.14.0-3vyhhypsq43ie2zjpp634apu6eqcucts/lib:/u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/gcc-
             4.8.5/hdf5-1.10.2-247ex6z2m4ehrczv6r7gibf3to2zsa5f/lib:/u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/gcc-4.8.5/zlib-1.2.11-4v3ticyykh2xcgw5fzltjrfuxkx7
             ipqh/lib:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/common/auxiliarySoftware/SuiteSparse/src:/u/danfitz/spack/opt/spack/l
             inux-rhel7-ppc64le/xl_r-16.1/mumps-5.1.1-3ivwblkwww3x3geewzye2v5swtucrtsv/lib:/u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/netlib-scalapack-
             2.0.2-r6jp3lwxdyceztqrux4zwlujuk3lb4ji/lib:/u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/suite-sparse-5.2.0-dmch5hznwgrmg3tcnfhmojixpzkncf26/
             lib:/u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/parmetis-4.0.3-ripisfb4stbqj5ueqswwgm3pkccppoir/lib:/u/danfitz/spack/opt/spack/linux-rhel7-
             ppc64le/xl_r-16.1/metis-5.1.0-wjzuwozzzsscsyqfwjrexn6kaet7rbha/lib:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/triutils/sr
             c:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/epetra/src:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/
             packages/teuchos/kokkoscomm/src:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/teuchos/kokkoscompat/src:/tmp/danfitz/spack-st
             age/spack-stage-AO_5dB/Trilinos/spack-build/packages/teuchos/remainder/src:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/teu
             chos/numerics/src:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/teuchos/comm/src:/tmp/danfitz/spack-stage/spack-stage-AO_5dB
             /Trilinos/spack-build/packages/teuchos/parameterlist/src:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/teuchos/parser/src:/u
             /danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/netlib-lapack-3.8.0-gyxkcm34uk2mx2sp2tkajit5qjcefqp4/lib64:/opt/ibmmath/essl/6.1/lib64:/tmp/danfi
             tz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/teuchos/core/src:/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packag
             es/kokkos/core/src: ../../amesos/src/libamesos.so.12.13 ../../aztecoo/src/libaztecoo.so.12.13 ../../epetraext/src/libepetraext.so.12.13 /u/danfitz/spack
             /opt/spack/linux-rhel7-ppc64le/xl_r-16.1/hypre-2.14.0-3vyhhypsq43ie2zjpp634apu6eqcucts/lib/libHYPRE.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/gc
             c-4.8.5/hdf5-1.10.2-247ex6z2m4ehrczv6r7gibf3to2zsa5f/lib/libhdf5.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/gcc-4.8.5/zlib-1.2.11-4v3ticyykh2xcgw
             5fzltjrfuxkx7ipqh/lib/libz.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/gcc-4.8.5/hdf5-1.10.2-247ex6z2m4ehrczv6r7gibf3to2zsa5f/lib/libhdf5_hl.so ..
             /../common/auxiliarySoftware/SuiteSparse/src/libtrilinosss.so.12.13 -lm /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/mumps-5.1.1-3ivwblkwww3
             x3geewzye2v5swtucrtsv/lib/libdmumps.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/mumps-5.1.1-3ivwblkwww3x3geewzye2v5swtucrtsv/lib/libmump
             s_common.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/mumps-5.1.1-3ivwblkwww3x3geewzye2v5swtucrtsv/lib/libpord.so /u/danfitz/spack/opt/sp
             ack/linux-rhel7-ppc64le/xl_r-16.1/netlib-scalapack-2.0.2-r6jp3lwxdyceztqrux4zwlujuk3lb4ji/lib/libscalapack.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc
             64le/xl_r-16.1/suite-sparse-5.2.0-dmch5hznwgrmg3tcnfhmojixpzkncf26/lib/libumfpack.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/suite-spar
             se-5.2.0-dmch5hznwgrmg3tcnfhmojixpzkncf26/lib/libamd.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/suite-sparse-5.2.0-dmch5hznwgrmg3tcnfhm
             ojixpzkncf26/lib/libcolamd.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/suite-sparse-5.2.0-dmch5hznwgrmg3tcnfhmojixpzkncf26/lib/libcholmo
             d.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/suite-sparse-5.2.0-dmch5hznwgrmg3tcnfhmojixpzkncf26/lib/libsuitesparseconfig.so /u/danfitz
             /spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/superlu-dist-5.2.2-duvrseth6foutnu24gq6mk6n3bpto636/lib/libsuperlu_dist.a /u/danfitz/spack/opt/spack/linu
             x-rhel7-ppc64le/xl_r-16.1/parmetis-4.0.3-ripisfb4stbqj5ueqswwgm3pkccppoir/lib/libparmetis.so /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/me
             tis-5.1.0-wjzuwozzzsscsyqfwjrexn6kaet7rbha/lib/libmetis.so ../../triutils/src/libtriutils.so.12.13 ../../epetra/src/libepetra.so.12.13 ../../teuchos/kok
             koscomm/src/libteuchoskokkoscomm.so.12.13 ../../teuchos/kokkoscompat/src/libteuchoskokkoscompat.so.12.13 ../../teuchos/remainder/src/libteuchosremainder
             .so.12.13 ../../teuchos/numerics/src/libteuchosnumerics.so.12.13 ../../teuchos/comm/src/libteuchoscomm.so.12.13 ../../teuchos/parameterlist/src/libteuch
             osparameterlist.so.12.13 ../../teuchos/parser/src/libteuchosparser.so.12.13 /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/netlib-lapack-3.8.0
             -gyxkcm34uk2mx2sp2tkajit5qjcefqp4/lib64/liblapack.so /opt/ibmmath/essl/6.1/lib64/libesslsmp.so ../../teuchos/core/src/libteuchoscore.so.12.13 ../../kokk
             os/core/src/libkokkoscore.so.12.13 /usr/lib64/libdl.so
     8541    cd /tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build/packages/ifpack/src && /u/danfitz/spack/opt/spack/linux-rhel7-ppc64le/xl_r-16.1/cmak
             e-3.11.3-4joy3z55k5hm6j75ssjougditfpsjh5j/bin/cmake -E cmake_symlink_library libifpack.so.12.13 libifpack.so.12 libifpack.so
     8542    make[2]: Leaving directory `/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build'
     8543    [ 84%] Built target ifpack
     8544    In file included from /u/danfitz/spack/var/spack/stage/trilinos-12.13.1-4kl2xjobtvkeiimmeunht447jfdolep2/Trilinos/packages/tpetra/core/src/Tpetra_CrsGra
             ph_SerialWrapperNode.cpp:54:
  >> 8545    /u/danfitz/spack/var/spack/stage/trilinos-12.13.1-4kl2xjobtvkeiimmeunht447jfdolep2/Trilinos/packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp:651:3: erro
             r: 1540-2206 Class with virtual base class and delegating constructor is not supported.
     8546      CrsGraph<LocalOrdinal, GlobalOrdinal, Node>::
     8547      ^
     8548    1 error generated.
  >> 8549    Error while processing /u/danfitz/spack/var/spack/stage/trilinos-12.13.1-4kl2xjobtvkeiimmeunht447jfdolep2/Trilinos/packages/tpetra/core/src/Tpetra_CrsGr
             aph_SerialWrapperNode.cpp.
  >> 8550    make[2]: *** [packages/tpetra/core/src/CMakeFiles/tpetra.dir/Tpetra_CrsGraph_SerialWrapperNode.cpp.o] Error 1
     8551    make[2]: *** Waiting for unfinished jobs....
     8552    make[2]: Leaving directory `/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build'
  >> 8553    make[1]: *** [packages/tpetra/core/src/CMakeFiles/tpetra.dir/all] Error 2
     8554    make[1]: Leaving directory `/tmp/danfitz/spack-stage/spack-stage-AO_5dB/Trilinos/spack-build'
  >> 8555    make: *** [all] Error 2
```

**Comment by mhoemmen** [comment]
@djfitzgerald Let's see if we can write a short reproducer.  The following code compiles with GCC 4.8.4 with the `-std=c++11` flag.  Does it compile with xlC?  
```
#include <iostream>
#include <memory>
#include <sstream>
#include <string>

class A {
public:
    virtual void foo () {
        std::cout << "A::foo" << std::endl;
    }
};

int stringToInt (const std::string& s) {
    std::istringstream is (s);
    int val = 0;
    is >> val;
    return val;
}

class B : public A {
public:
    B (const int x) : x_ (x)
    {
        std::cout << "B constructor" << std::endl;
    }

    // Delegating constructor
    B (const std::string& s) : B(stringToInt (s))
    {}

    // overrides base class virtual method
    void foo () override {
        std::cout << "B::foo" << std::endl;
    }

private:
    int x_;
};

int main (int argc, char* argv[]) {
    std::unique_ptr<A> a0 (new A);
    std::unique_ptr<A> b0 (new B (42));
    std::unique_ptr<A> b1 (new B (std::string ("43")));

    a0->foo ();
    b0->foo ();
    b1->foo ();

    return 0;
}
```

**Comment by mhoemmen** [comment]
@nmhamster We're trying to come up w/ a short reproducer to see if this is an xlC issue.

**Comment by djfitzgerald** [comment]
@mhoemmen Yes, I can get this to compile with xlc if I use `-std=c++11`

**Comment by mhoemmen** [comment]
@djfitzgerald Could there be something wrong with Spack's Trilinos configuration?  Do you have a stand-alone (non-Spack) Trilinos configuration to try?  If not, we could help you set one up.  If Spack is broken, we should push this to the Spack developers.

**Comment by djfitzgerald** [comment]
@mhoemmen I do not have a non-Spack Trilinos configuration to try.  I'll go about trying to set one up, but would glady accept any help you have to offer.

**Comment by mhoemmen** [comment]
@djfitzgerald Just for now -- would it be possible to find out what arguments Spack gives to CMake?  If not, could you please e-mail me the `CMakeCache.txt` file in your build directory?  Thanks!

**Comment by djfitzgerald** [comment]
@mhoemmen Dumb question, but what's your email?

**Comment by mhoemmen** [comment]
@djfitzgerald Obfuscation: remove the n, add sandia's domain

**Comment by mhoemmen** [comment]
(I deleted my comment, since you have my e-mail address now; thanks!)

**Comment by nmhamster** [comment]
@mhoemmen - new XL on its way. FYI.

**Comment by mhoemmen** [comment]
@nmhamster Thanks! :-D  We're just trying to figure out if this is a compiler or a configuration issue.

**Comment by djfitzgerald** [comment]
I've been unable to get a non-Spack based Trilinos installed and configured.  The best I can do after quite a bit of work is get install 58% of the way.  I've really run out of time to work on this, and am unsure how to proceed.

**Comment by mhoemmen** [comment]
@nmhamster , @trilinos/framework , and/or @mwglass may be able to help.

I'm swamped this week but if this becomes a priority, I could work with you or other IBM folks on getting a non-Spack Trilinos configuration.

**Comment by djfitzgerald** [comment]
This is unfortunately not going to become a priority on our end.  As far as our team is concerned, we've got the fix that we need committed into Spack, and being able to push it upstream into Trilinos would just be gravy.  Maybe if I can get some relief with #2781 I can follow this through.

**Comment by hubert-reinterpretcast** [comment]
Reproducer for the issue:
```
struct A { };
struct B : virtual A {
  B(int, const B &);
};
B::B(int, const B &b) : B(b) { }
```

The error message from the XL compiler is indicating some issue in the implementation, and not an error in the source code.

**Comment by djfitzgerald** [comment]
@mhoemmen Of that I know nothing about; for that you'd have to talk to someone from the XL Compilers team.  I'm just a humble end-user, much as yourself :-P

**Comment by hubert-reinterpretcast** [comment]
> The error message from the XL compiler is indicating some issue in the implementation, and not an error in the source code.

Compiler support has been updated in Beta 3 of XL C/C++ for Linux V16.1.1. I have verified that the reproducer from my comment above now compiles with IBM xlC.

@mhoemmen If you do not have access to the beta; information on joining the beta program is available at https://ibm.biz/join-xl-beta. I will send @djfitzgerald information on how to access the beta internally.
